File: examples/mcp-redact/redact-tool-call.mjs
/**
 * MCP tool calls through the AI-context boundary (issue #327, migrated onto
 * `@redact-secret/adapter-ai-context` by redact-secret-adapters#12): given a
 * boundary, sanitizes a `CallToolResult`'s content blocks and
 * `structuredContent`, or a tool call's argument object, and returns the
 * boundary's own outcome shape (`docs/reference/ai-context-boundary.md`):
 *
 * - `{ outcome: "ok", value, findings }`: `redact` findings replaced,
 *   `warn`/`allow` text unchanged; `value` is the only safe thing to return.
 * - `{ outcome: "blocked", reason, code? }`: any `block` finding, a key that
 *   would be redacted, a limit, a non-JSON value, or a core failure, anywhere
 *   in the call. No value, no findings, nothing derived from input. The caller
 *   returns {@link buildBlockedResult} instead: MCP has a first-class "tool
 *   error" outcome (`CallToolResult.isError`), so a blocked call maps onto
 *   that rather than a partial result.
 * - `{ outcome: "aborted" }`: the signal fired.
 *
 * This file owns only the MCP shape: which content blocks carry text, and
 * JSON-in-text. Every scan, every traversal, every limit and every failure
 * mapping is the adapter's. There is no walker, marker or leaf budget here.
 *
 * No `@modelcontextprotocol/sdk` import: `CallToolResult` and its content
 * blocks are a structural (duck-typed) shape here.
 */

/**
 * The most content blocks one result may carry. Past it the whole result is
 * blocked as `limit_exceeded`; blocks are never dropped or passed on unscanned.
 */
export const DEFAULT_MAX_CONTENT_BLOCKS = 200;

/** Fixed, input-free text for every blocked outcome. Never carries the
 * input, a matched value, or the underlying error's own message. */
export const BLOCKED_MESSAGE =
  "This MCP tool call was blocked by secret-redaction policy. No content, arguments, or error detail is included.";

function isPlainObject(value) {
  if (typeof value !== "object" || value === null) return false;
  const proto = Object.getPrototypeOf(value);
  return proto === Object.prototype || proto === null;
}

function requireBoundary(boundary, caller) {
  if (boundary === null || typeof boundary !== "object" || typeof boundary.sanitizeValue !== "function") {
    throw new TypeError(`${caller}: boundary must be an @redact-secret/adapter-ai-context boundary`);
  }
}

const UNSUPPORTED = Object.freeze({ outcome: "blocked", reason: "unsupported_value" });
const OVER_LIMIT = Object.freeze({ outcome: "blocked", reason: "limit_exceeded" });

/**
 * One piece of text that may be a JSON-serialized object or array (an MCP
 * "JSON-in-text" result, such as a stringified API response). When it
 * parses as one, the parsed value goes through `sanitizeValue` and is
 * re-serialized, which, unlike scanning the raw JSON text as one opaque
 * string, can never place a placeholder outside a quoted string and corrupt
 * the JSON. Anything else is scanned as text.
 */
function sanitizeTextOrJson(boundary, text, options) {
  let parsed;
  try {
    parsed = JSON.parse(text);
  } catch {
    parsed = undefined;
  }
  if (Array.isArray(parsed) || isPlainObject(parsed)) {
    const outcome = boundary.sanitizeValue(parsed, options);
    return outcome.outcome === "ok" ? { ...outcome, value: JSON.stringify(outcome.value) } : outcome;
  }
  return boundary.sanitizeText(text, options);
}

/**
 * Text blocks and embedded text resources are scanned. `image`, `audio`,
 * `resource_link`, and embedded blob resources (base64 binary) pass through
 * unchanged: non-text content is outside the boundary contract, a documented
 * false negative.
 */
function sanitizeContentBlock(boundary, block, options) {
  if (!isPlainObject(block)) return UNSUPPORTED;
  if (block.type === "text" && typeof block.text === "string") {
    const outcome = sanitizeTextOrJson(boundary, block.text, options);
    return outcome.outcome === "ok" ? { ...outcome, value: { ...block, text: outcome.value } } : outcome;
  }
  if (block.type === "resource" && isPlainObject(block.resource) && typeof block.resource.text === "string") {
    const outcome = sanitizeTextOrJson(boundary, block.resource.text, options);
    return outcome.outcome === "ok"
      ? { ...outcome, value: { ...block, resource: { ...block.resource, text: outcome.value } } }
      : outcome;
  }
  return { outcome: "ok", value: block, findings: [] };
}

/** The fixed `CallToolResult`-shaped tool error every blocked or aborted outcome maps
 * to: no content, no arguments, no error detail beyond the fixed message. */
export function buildBlockedResult() {
  return { content: [{ type: "text", text: BLOCKED_MESSAGE }], isError: true };
}

/**
 * Sanitizes a tool call's argument object (MCP
 * `CallToolRequest.params.arguments`, a plain object tree) with
 * `sanitizeValue`: every string value and every key is scanned. On a
 * non-`ok` outcome the caller must not forward the original arguments.
 *
 * @param {import("@redact-secret/adapter-ai-context").AiContextBoundary} boundary
 * @param {unknown} args
 * @param {{ signal?: AbortSignal }} [options]
 */
export function redactArguments(boundary, args, { signal } = {}) {
  requireBoundary(boundary, "redactArguments");
  if (!isPlainObject(args)) return UNSUPPORTED;
  // The contract has no label for tool arguments yet (#612 defines the MCP
  // boundary); they cross into a tool, so they are labelled `context`.
  return boundary.sanitizeValue(args, { boundary: "context", signal });
}

/**
 * Sanitizes a `CallToolResult`'s `content` blocks and `structuredContent`
 * before it reaches the client, a log, or model context. All or nothing:
 * the first non-`ok` block, or a non-`ok` `structuredContent`, is the
 * outcome for the whole result. Other top-level fields (`isError`, `_meta`)
 * are copied unchanged.
 *
 * @param {import("@redact-secret/adapter-ai-context").AiContextBoundary} boundary
 * @param {unknown} result
 * @param {{ signal?: AbortSignal, maxContentBlocks?: number }} [options]
 */
export function redactToolResult(boundary, result, { signal, maxContentBlocks = DEFAULT_MAX_CONTENT_BLOCKS } = {}) {
  requireBoundary(boundary, "redactToolResult");
  if (signal?.aborted === true) return Object.freeze({ outcome: "aborted" });
  if (!isPlainObject(result)) return UNSUPPORTED;
  if (result.content !== undefined && !Array.isArray(result.content)) return UNSUPPORTED;
  const contentIn = result.content ?? [];
  if (contentIn.length > maxContentBlocks) return OVER_LIMIT;

  const options = { boundary: "tool-result", signal };
  const findings = [];
  const content = [];
  for (const block of contentIn) {
    const outcome = sanitizeContentBlock(boundary, block, options);
    if (outcome.outcome !== "ok") return outcome;
    findings.push(...outcome.findings);
    content.push(outcome.value);
  }

  const out = { ...result, content };
  if (result.structuredContent !== undefined) {
    const outcome = boundary.sanitizeValue(result.structuredContent, options);
    if (outcome.outcome !== "ok") return outcome;
    findings.push(...outcome.findings);
    out.structuredContent = outcome.value;
  }
  return Object.freeze({ outcome: "ok", value: out, findings: Object.freeze(findings) });
}

File: examples/mcp-redact/python/redact_tool_call.py
"""The pure redaction core for MCP tool calls (issue #327): given an
injected ``scan_and_redact``, redacts a ``CallToolResult``'s content blocks
/ structured content, or a tool call's argument dict, and maps the outcome
onto the issue's policy contract:

- ``redact`` findings: the span is replaced, the call continues.
- ``warn`` findings: the text passes through unchanged; still reported.
- ``block`` findings, or a ``scan_and_redact`` failure: the *whole* call
  becomes a fixed, input-free block outcome -- never a partial or per-leaf
  marker. This differs from ``examples/tracing-masking``'s per-leaf
  ``BLOCK_MARKER``: MCP already has a first-class "tool error" outcome
  (``CallToolResult.is_error``), so blocking maps onto that instead of a
  leaf-level placeholder.

No ``mcp``/``fastmcp`` import: ``CallToolResult`` and its content blocks are
a structural (duck-typed) shape here -- a plain dict tree -- the same choice
``examples/tracing-masking/python/redact_span_attributes.py`` makes for
OpenTelemetry's ``SpanProcessor``. This file is testable without the built
native extension or an MCP SDK installed.

Mirrors ``../redact-tool-call.mjs`` field for field; see that file's
docstring for the design rationale.
"""

from __future__ import annotations

import json
from typing import Any, Callable, Optional

__all__ = [
    "DEFAULT_LIMITS",
    "BLOCKED_MESSAGE",
    "LIMIT_MARKER",
    "CYCLE_MARKER",
    "build_blocked_result",
    "redact_arguments",
    "redact_tool_result",
]

# Bounds enforced while walking arguments and JSON-in-text content. Beyond
# these, a subtree is marked with LIMIT_MARKER and dropped rather than
# passed through unmasked -- the call itself is not blocked, since these are
# complexity guards, not secret-detection outcomes.
DEFAULT_LIMITS: dict[str, int] = {
    "max_depth": 8,
    "max_array_length": 1000,
    "max_object_keys": 200,
    "max_string_length": 200_000,
    "max_total_leaves": 5000,
    "max_content_blocks": 200,
}

# Fixed, input-free text for every blocked outcome. Never carries the input,
# a matched value, or the underlying error's own message.
BLOCKED_MESSAGE = (
    "This MCP tool call was blocked by secret-redaction policy. "
    "No content, arguments, or error detail is included."
)

LIMIT_MARKER = "[REDACTED:LIMIT_EXCEEDED]"
CYCLE_MARKER = "[REDACTED:CYCLE]"


def _is_plain_dict(value: Any) -> bool:
    return type(value) is dict


def _is_plain_list(value: Any) -> bool:
    return type(value) is list


class _Context:
    __slots__ = ("policy", "limits", "budget", "blocked", "block_reason", "findings")

    def __init__(self, policy: Any, limits: dict[str, int]) -> None:
        self.policy = policy
        self.limits = limits
        self.budget = {"leaves": limits["max_total_leaves"]}
        self.blocked = False
        self.block_reason: Optional[str] = None
        self.findings: list[Any] = []


def _make_context(scan_and_redact: Callable[..., Any], policy: Any, limits: Optional[dict[str, int]]) -> _Context:
    if not callable(scan_and_redact):
        raise TypeError("scan_and_redact must be callable")
    merged_limits = {**DEFAULT_LIMITS, **(limits or {})}
    return _Context(policy, merged_limits)


def _redact_leaf(scan_and_redact: Callable[..., Any], text: str, ctx: _Context) -> str:
    if len(text) > ctx.limits["max_string_length"]:
        return LIMIT_MARKER
    if ctx.budget["leaves"] <= 0:
        return LIMIT_MARKER
    ctx.budget["leaves"] -= 1

    try:
        result = scan_and_redact(text, policy=ctx.policy)
    except Exception:
        ctx.blocked = True
        ctx.block_reason = "core_error"
        return ""

    ctx.findings.extend(result.findings)
    if any(finding.action == "block" for finding in result.findings):
        ctx.blocked = True
        ctx.block_reason = "policy"
        return ""
    return result.text


def _redact_json_value(scan_and_redact: Callable[..., Any], value: Any, ctx: _Context, depth: int, seen: set) -> Any:
    if ctx.blocked:
        return value

    if isinstance(value, str):
        return _redact_leaf(scan_and_redact, value, ctx)

    if _is_plain_list(value):
        if depth >= ctx.limits["max_depth"]:
            return LIMIT_MARKER
        if id(value) in seen:
            return CYCLE_MARKER
        seen.add(id(value))
        try:
            bounded = value[: ctx.limits["max_array_length"]]
            return [_redact_json_value(scan_and_redact, item, ctx, depth + 1, seen) for item in bounded]
        finally:
            seen.discard(id(value))

    if _is_plain_dict(value):
        if depth >= ctx.limits["max_depth"]:
            return LIMIT_MARKER
        if id(value) in seen:
            return CYCLE_MARKER
        seen.add(id(value))
        try:
            keys = list(value.keys())[: ctx.limits["max_object_keys"]]
            return {key: _redact_json_value(scan_and_redact, value[key], ctx, depth + 1, seen) for key in keys}
        finally:
            seen.discard(id(value))

    # Numbers, booleans, None, and non-plain objects are left unchanged:
    # only plain dicts, lists, and strings are walked.
    return value


def _redact_text_or_json(scan_and_redact: Callable[..., Any], text: str, ctx: _Context) -> str:
    """Redacts one piece of text that may be a JSON-serialized object/array
    (an MCP "JSON-in-text" result). When it parses as a JSON object or
    array, every string leaf inside is redacted and the value is
    re-serialized with JS-compatible compact separators (``","``/``":"``,
    no spaces) so JS and Python produce byte-identical output. Anything
    else is scanned as opaque text.
    """
    if ctx.blocked:
        return text
    try:
        parsed = json.loads(text)
    except ValueError:
        parsed = None

    if _is_plain_list(parsed) or _is_plain_dict(parsed):
        redacted = _redact_json_value(scan_and_redact, parsed, ctx, 0, set())
        if ctx.blocked:
            return ""
        return json.dumps(redacted, separators=(",", ":"))
    return _redact_leaf(scan_and_redact, text, ctx)


def _redact_content_block(scan_and_redact: Callable[..., Any], block: Any, ctx: _Context) -> Any:
    if ctx.blocked:
        return block
    if not _is_plain_dict(block):
        return block

    if block.get("type") == "text" and isinstance(block.get("text"), str):
        return {**block, "text": _redact_text_or_json(scan_and_redact, block["text"], ctx)}

    # An embedded text resource (`{"type": "resource", "resource": {"text": ...}}`)
    # carries scannable text; a blob resource (`resource.blob`) is base64
    # binary and, like `image`/`audio`/`resource_link` blocks, passes
    # through unchanged -- the documented non-text false-negative boundary.
    resource = block.get("resource")
    if block.get("type") == "resource" and _is_plain_dict(resource) and isinstance(resource.get("text"), str):
        return {
            **block,
            "resource": {**resource, "text": _redact_text_or_json(scan_and_redact, resource["text"], ctx)},
        }

    return block


def build_blocked_result() -> dict[str, Any]:
    """The fixed ``CallToolResult``-shaped tool error every blocked outcome
    maps to: no content, no arguments, no error detail beyond the fixed
    message."""
    return {"content": [{"type": "text", "text": BLOCKED_MESSAGE}], "isError": True}


def redact_arguments(
    scan_and_redact: Callable[..., Any],
    args: dict[str, Any],
    *,
    policy: Optional[Any] = None,
    limits: Optional[dict[str, int]] = None,
) -> dict[str, Any]:
    """Redacts a tool call's argument dict. Returns ``{"outcome": "blocked",
    ...}`` -- with no ``"arguments"`` key -- when any argument value
    contains a ``block`` finding or ``scan_and_redact`` fails; the caller
    must not forward the original arguments to the tool in that case.
    """
    ctx = _make_context(scan_and_redact, policy, limits)
    if not _is_plain_dict(args):
        raise TypeError("redact_arguments: args must be a plain dict")
    redacted = _redact_json_value(scan_and_redact, args, ctx, 0, set())
    if ctx.blocked:
        return {"outcome": "blocked", "blockReason": ctx.block_reason, "findings": ctx.findings}
    return {"outcome": "ok", "arguments": redacted, "findings": ctx.findings}


def redact_tool_result(
    scan_and_redact: Callable[..., Any],
    result: dict[str, Any],
    *,
    policy: Optional[Any] = None,
    limits: Optional[dict[str, int]] = None,
) -> dict[str, Any]:
    """Redacts a ``CallToolResult``'s ``content`` blocks and
    ``structuredContent`` before it reaches the client or model context.
    Returns ``{"outcome": "blocked", ...}`` -- with no ``"result"`` key --
    on any ``block`` finding or core failure; the caller must return
    ``build_blocked_result()`` instead of the original result in that case.
    """
    ctx = _make_context(scan_and_redact, policy, limits)
    if not _is_plain_dict(result):
        raise TypeError("redact_tool_result: result must be a plain dict")

    content_in = result.get("content") if _is_plain_list(result.get("content")) else []
    bounded = content_in[: ctx.limits["max_content_blocks"]]
    content = [_redact_content_block(scan_and_redact, block, ctx) for block in bounded]

    has_structured_content = "structuredContent" in result
    structured_content = result.get("structuredContent")
    if not ctx.blocked and _is_plain_dict(structured_content):
        structured_content = _redact_json_value(scan_and_redact, structured_content, ctx, 0, set())

    if ctx.blocked:
        return {"outcome": "blocked", "blockReason": ctx.block_reason, "findings": ctx.findings}

    out = {**result, "content": content}
    if has_structured_content:
        out["structuredContent"] = structured_content
    return {"outcome": "ok", "result": out, "findings": ctx.findings}
